### PR TITLE
Refine JITServer AOT cache flag setting

### PR DESCRIPTION
**Pull request by cjjdespres**
When using the JITServer AOT cache and ignoring the local SCC, portable AOT is now enabled by default only if the JVM is running in a container, and only if there is no explicit `-XX:[+|-]PortableSharedCache` option present on the command line.

The `-XX:[+|-]PortableSharedCache` options are also now honoured when using the JITServer AOT cache (having the effect of explicitly enabling or disabling portable AOT) even if the SCC is disabled.

Related: https://github.com/eclipse-openj9/openj9/issues/18990

### CODE REVIEW

**Comment by cjjdespres** [comment]
Attn @mpirvu, @AlexeyKhrabrov. I think (and correct me if I'm wrong) that it's still desirable to have portable AOT enabled by default for AOT cache clients in the same circumstances that it is enabled by default for local AOT. This PR does this, and provides a way of explicitly overriding this default.

If this is still too much, I can take a look again at the few header incompatibilities that prevented code sharing between clients in containers running with and without a local SCC, and see if I can set the relevant JVM flags/features more precisely.

**Comment by AlexeyKhrabrov** [comment]
> I think (and correct me if I'm wrong) that it's still desirable to have portable AOT enabled by default for AOT cache clients in the same circumstances that it is enabled by default for local AOT.

That's the part I disagree with. My understanding is that the whole point of the portable AOT mode is to handle execution environment differences between where the SCC is originally created and populated (an intermediate container during image build), and the eventual deployed container where the application will run. In practice, these two environment are most likely to be different, which is why enabling portable AOT by default makes sense. The alternative would be to maintain multiple images for different environment configurations, which would be difficult to manage.

With the AOT cache, the multi-configuration issue is handled transparently, and the only advantage of the portable AOT mode that I can think of is sharing server-cached methods across heterogeneous client JVM environments, in which case it's a trade-off between compiled code performance and the AOT cache hit rate. Whereas In the mostly homogeneous case (in the sense that for each distinct configuration, there are still many clients that share it), which I think is more common in practice, portable AOT has no advantages and can only degrade performance, which is why I don't think it should be enabled by default.

But regardless of what the right default is, this PR is definitely an improvement over the current situation.

File: runtime/vm/jvminit.c [hunk 1]
```diff
@@ -2133,27 +2138,33 @@ VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved)
...
 				}
-#if (defined(J9VM_ARCH_X86) || defined(J9VM_ARCH_S390) || defined(J9VM_ARCH_POWER) || defined(J9VM_ARCH_AARCH64)) && defined(J9VM_OPT_JITSERVER)
-				// Regardless of whether or not the local SCC is enabled, we still need to enable portable AOT if
-				// the options suggest we're going to connect to a JITServer AOT cache.
-				if (argIndexXXJITServerAOTCache > argIndexXXNoJITServerAOTCache) {
-					vm->extendedRuntimeFlags2 |= J9_EXTENDED_RUNTIME2_ENABLE_PORTABLE_SHARED_CACHE;
+#if defined(J9VM_OPT_JITSERVER)
+				if ((argIndexXXJITServerAOTCache > argIndexXXNoJITServerAOTCache) &&
+				    (argIndexXXJITServerAOTCacheIgnoreLocalSCC > argIndexXXNoJITServerAOTCacheIgnoreLocalSCC)) {
+					// If we are using the JITServer AOT cache and ignoring the local SCC (which we detect by checking
```
Comment (mpirvu) [inline runtime/vm/jvminit.c:2148]: This is a .c file and it should use c style comments not c++.
Also, I think that VM coding style wants the `&&` to be at the beginning of the next line.

**Review by mpirvu — APPROVED** [review]
I only have a small suggestion.
At some point we need to run some benchmarks to see the performance effect of portable AOT. Also, if possible, connect two clients with distinct configurations (maybe one using compressedrefs shift 1 and the other one 3) and see the increase in footprint at the server when not using portable AOT.

File: runtime/vm/jvminit.c (hunk 1 above)
Comment (cjjdespres) [reply runtime/vm/jvminit.c]: Fixed.

**Review by mpirvu — APPROVED** [review]

**Comment by mpirvu** [comment]
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit,alinux64jit jdk17

**Comment by mpirvu** [comment]
Failure:
plinux - testJitserverArguments_0
```
4:25:46  Testing: Test SSL success condition
14:25:46  Test start time: 2024/04/02 14:25:45 Eastern Standard Time
14:25:46  Running command: bash /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_ppc64le_linux_jit_Personal_testList_0/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/jitserver/jitserverScript.sh /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_ppc64le_linux_jit_Personal_testList_0/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/jitserver /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_ppc64le_linux_jit_Personal_testList_0/jdkbinary/j2sdk-image/bin "-Xjit" "-XX:+UseJITServer -Xjit:count=0,verbose={JITServer},verbose={JITServerConns},verbose={compilePerformance} -XX:-JITServerLocalSyncCompiles -XX:JITServerSSLRootCerts=cert.pem" false false
14:25:46  Time spent starting: 3 milliseconds
14:25:48  Time spent executing: 2488 milliseconds
14:25:48  Test result: FAILED
14:25:48  Output from test:
14:25:48   [OUT] start running script
14:25:48   [OUT] Generate SSL certificates
14:25:48   [OUT] Creating SSL certificates
14:25:48   [OUT] Certificates generated
14:25:48   [OUT] Starting /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_ppc64le_linux_jit_Personal_testList_0/jdkbinary/j2sdk-image/bin/jitserver -XX:JITServerPort=45848 -XX:JITServerHealthProbePort=47372  -Xjit -XX:JITServerSSLKey=key.pem -XX:JITServerSSLCert=cert.pem
14:25:48   [OUT] JITSERVER DOES NOT EXIST
14:25:48   [OUT] finished script
14:25:48   [ERR] Generating RSA private key, 2048 bit long modulus (2 primes)
14:25:48   [ERR] ...................+++++
14:25:48   [ERR] ..................................................+++++
14:25:48   [ERR] e is 65537 (0x010001)
14:25:48   [ERR] Generating a RSA private key
14:25:48   [ERR] ..................+++++
14:25:48   [ERR] ............................+++++
14:25:48   [ERR] writing new private key to 'wrongKey.pem'
14:25:48   [ERR] -----
14:25:48   [ERR] 
14:25:48   [ERR] JITServer is ready to accept incoming requests
14:25:48   [ERR] can't bind server address: Address already in use
14:25:48   [ERR] Failed to open health socket on port 47372
```

zLinux: cmdLineTester_criu_nonPortableRestore_6
```
14:11:08  Running test cmdLineTester_criu_nonPortableRestore_6 ...
14:11:08  ===============================================
14:11:08  cmdLineTester_criu_nonPortableRestore_6 Start Time: Tue Apr  2 18:11:07 2024 Epoch Time (ms): 1712081467358
14:11:08  variation: -Xgcpolicy:optavgpause
14:11:08  JVM_OPTIONS: -XX:+UseJITServer -Xgcpolicy:optavgpause 
...
4:13:14  Testing: Create and Restore Criu Checkpoint Image once - MethodTypeDeadlockTest
14:13:14  Test start time: 2024/04/02 18:13:13 Coordinated Universal Time
14:13:14  Running command: bash /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_s390x_linux_jit_Personal_testList_1/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu/criuScript.sh /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_s390x_linux_jit_Personal_testList_1/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_s390x_linux_jit_Personal_testList_1/jdkbinary/j2sdk-image/bin/java "-XX:+UseJITServer -Xgcpolicy:optavgpause  -XX:+ThrowOnDelayedCheckpointOperation -XX:sleepMillisecondsForNotCheckpointSafe=20 -Xtrace:print=j9vm.731 --add-opens java.base/jdk.internal.misc=ALL-UNNAMED  --add-exports java.base/openj9.internal.criu=ALL-UNNAMED" org.openj9.criu.DeadlockTest MethodTypeDeadlockTest 1 false false
14:13:14  Time spent starting: 7 milliseconds
14:13:19  Time spent executing: 4025 milliseconds
14:13:19  Test result: FAILED
14:13:19  Output from test:
14:13:19   [OUT] start running script
14:13:19   [OUT] export GLIBC_TUNABLES=glibc.cpu.hwcaps=-XSAVEC,-XSAVE,-AVX2,-ERMS,-AVX,-AVX_Fast_Unaligned_Load
14:13:19   [OUT] export LD_BIND_NOT=on
14:13:19   [OUT] /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_s390x_linux_jit_Personal_testList_1/jdkbinary/j2sdk-image/bin/java -XX:+EnableCRIUSupport -XX:+UseJITServer -Xgcpolicy:optavgpause  -XX:+ThrowOnDelayedCheckpointOperation -XX:sleepMillisecondsForNotCheckpointSafe=20 -Xtrace:print=j9vm.731 --add-opens java.base/jdk.internal.misc=ALL-UNNAMED  --add-exports java.base/openj9.internal.criu=ALL-UNNAMED -cp /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_s390x_linux_jit_Personal_testList_1/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu/criu.jar org.openj9.criu.DeadlockTest MethodTypeDeadlockTest 1
14:13:19   [OUT] Pre-checkpoint
14:13:19   [OUT] main: Tue Apr 02 18:13:14 UTC 2024, Performing CRIUSupport.checkpointJVM(), System.currentTimeMillis(): 1712081594091, System.nanoTime(): 1712081594087619911
14:13:19   [OUT] JVMJITM044W Some or all compiled code in the code cache invalidated post restore.
14:13:19   [OUT] initiate restore
14:13:19   [OUT] Error (criu/protobuf.c:72): Unexpected EOF on (empty-image)
14:13:19   [OUT] Error (criu/arch/s390/crtools.c:776): Cannot open core for virt pid 4073819: No such file or directory
14:13:19   [OUT] Error (criu/cr-restore.c:2142): Restoring regs for 4073819 failed: No such file or directory
14:13:19   [OUT] Error (criu/cr-restore.c:2520): Killing processes because of failure on restore.
14:13:19   [OUT] The Network was unlocked so some data or a connection may have been lost.
14:13:19   [OUT] Error (criu/cr-restore.c:2547): Restoring FAILED.
14:13:19   [OUT] Removed test output files
14:13:19   [OUT] finished script
```
alinux64 failed due to infra
```
hudson.remoting.ProxyException: java.io.IOException: No space left on device
```

**Comment by mpirvu** [comment]
Since the failures are not generated by these changes, I am going to merge this PR.

**Pull request merged**